core/node/proof_data_handler/src/tee_request_processor.rs

```diff
@@ -47,49 +51,52 @@ impl TeeRequestProcessor {
     ) -> Result<Option<Json<TeeProofGenerationDataResponse>>, RequestProcessorError> {
         tracing::info!("Received request for proof generation data: {:?}", request);
 
-        let mut min_batch_number = self.config.tee_config.first_tee_processed_batch;
-        let mut missing_range: Option<(L1BatchNumber, L1BatchNumber)> = None;
+        let batch_ignored_timeout = ChronoDuration::days(10);
+        let min_batch_number = self.config.tee_config.first_tee_processed_batch;
 
-        let result = loop {
-            let Some(l1_batch_number) = self
+        loop {
+            let Some(locked_batch) = self
                 .lock_batch_for_proving(request.tee_type, min_batch_number)
                 .await?
             else {
-                // No job available
-                return Ok(None);
+                return Ok(None); // no job available
             };
+            let batch_number = locked_batch.l1_batch_number;
 
             match self
-                .tee_verifier_input_for_existing_batch(l1_batch_number)
+                .tee_verifier_input_for_existing_batch(batch_number)
                 .await
             {
                 Ok(input) => {
                     break Ok(Some(Json(TeeProofGenerationDataResponse(Box::new(input)))));
                 }
                 Err(RequestProcessorError::ObjectStore(ObjectStoreError::KeyNotFound(_))) => {
-                    missing_range = match missing_range {
-                        Some((start, _)) => Some((start, l1_batch_number)),
-                        None => Some((l1_batch_number, l1_batch_number)),
+                    let duration = Utc::now().signed_duration_since(locked_batch.created_at);
+                    let status = if duration > batch_ignored_timeout {
+                        TeeProofGenerationJobStatus::PermanentlyIgnored
+                    } else {
+                        TeeProofGenerationJobStatus::Failed
                     };
-                    self.unlock_batch(l1_batch_number, request.tee_type).await?;
-                    min_batch_number = l1_batch_number + 1;
+                    self.unlock_batch(batch_number, request.tee_type, status)
+                        .await?;
+                    tracing::warn!(
+                        "Assigned status {} to batch {} created at {}",
+                        status,
+                        batch_number,
+                        locked_batch.created_at
+                    );
                 }
                 Err(err) => {
-                    self.unlock_batch(l1_batch_number, request.tee_type).await?;
+                    self.unlock_batch(
```

Comment:
Dumb question: Why does this unlocking work like that? Suppose this server loses connection to Postgres for a moment, resulting in a `RequestProcessorError::Dal` error. IIUC, the batch will be marked as unlocked here, but there's seemingly no reason to do so.

Reply:
Notice that the batch that failed due to `RequestProcessorError::Dal` or `RequestProcessorError::GeneralError` is guaranteed to be reconsidered by the SQL query that fetches the next batch to be TEE-proven:

https://github.com/matter-labs/zksync-era/blob/9188f8a3107f9a46d5b74c99efe7a1db5ccda56d/core/lib/dal/src/tee_proof_generation_dal.rs#L83-L84

Would you prefer me to implement explicit retry logic in case of `RequestProcessorError::Dal` errors?

Reply:
I see your point; I've misunderstood what `Failed` status means (I thought it means proof generation itself fails, but AFAIU it's also used for failures related to preparing inputs for proof generation). Regarding the query: Do I read it correctly that it uses the same timeout strategy that is used for force-unlocking batches? If so, why not use a smaller timeout or no timeout at all? Going with my previous example, IIUC, a transient network connectivity issue may lead to a batch getting ignored for 10 minutes (with the default config); this sounds inefficient.

Reply:
> (...) I've misunderstood what Failed status means (I thought it means proof generation itself fails, but AFAIU it's also used for failures related to preparing inputs for proof generation). (...)

I just refined the docs a bit to clear it up:
https://github.com/matter-labs/zksync-era/blob/0dd01210668c1fb25defa0cbe9723fc5ff3b1abd/core/lib/dal/src/tee_proof_generation_dal.rs#L32-L36

> (...) Regarding the query: Do I read it correctly that it uses the same timeout strategy that is used for force-unlocking batches? (...)

So, we are talking about two different timeouts here – I refined the docs for them as well:
https://github.com/matter-labs/zksync-era/blob/0dd01210668c1fb25defa0cbe9723fc5ff3b1abd/core/lib/config/src/configs/proof_data_handler.rs#L12-L17

> (...) If so, why not use a smaller timeout or no timeout at all? Going with my previous example, IIUC, a transient network connectivity issue may lead to a batch getting ignored for 10 minutes (with the default config); this sounds inefficient. (...)

The [timeout used in the SQL query](https://github.com/matter-labs/zksync-era/blob/9188f8a3107f9a46d5b74c99efe7a1db5ccda56d/core/lib/dal/src/tee_proof_generation_dal.rs#L84) is `tee_proof_generation_timeout_in_secs`, and if we set it to `0`, it means we keep retrying all [`Failed`](https://github.com/matter-labs/zksync-era/blob/0dd01210668c1fb25defa0cbe9723fc5ff3b1abd/core/lib/dal/src/tee_proof_generation_dal.rs#L36) batches immediately,  before proceeding with any other batches. If we have a batch where preparing inputs for proof generation always fails (for any reason), it would effectively block TEE proving progress. This happens because we would keep retrying until either the input preparation for that batch succeeds or the `tee_batch_permanently_ignored_timeout_in_hours` timeout expires.

Reply:
What I mean by "force-unlocking" is the transition between `picked_by_prover` to "unpicked" (which would be transformed to `picked_by_prover` again; it's not a real state). So, the question is whether it makes sense to make the timeout for this operation the same as the timeout for `failed` batches. IDK; intuitively, at least for some types of errors, the default timeout value of 10 minutes looks too large, but if a typical error takes this long to resolve, happy to roll with it.

Reply:
Gotcha. Decreased the retry interval for failed batches from 10 minutes to 1 minute [here](https://github.com/matter-labs/zksync-era/pull/2980/commits/ada62febcbf6c01e21ebdb0242823c4feb1303f5). ✅ Hopefully, that works better now.